File: erlang/rna-transcription/src/rna_transcription.erl
-module(rna_transcription).

-export([to_rna/1]).


to_rna(_Strand) -> [transcribe_nucleotide(X) || X <- _Strand].

transcribe_nucleotide($G) -> $C;
transcribe_nucleotide($C) -> $G;
transcribe_nucleotide($T) -> $A;
transcribe_nucleotide($A) -> $U.

File: elixir/rna-transcription/lib/rna_transcription.ex
defmodule RnaTranscription do
  @doc """
  Transcribes a character list representing DNA nucleotides to RNA

  ## Examples

    iex> RnaTranscription.to_rna(~c"ACTG")
    ~c"UGAC"
  """
  @spec to_rna([char]) :: [char]
  def to_rna(dna) do
    Enum.reverse(Enum.reduce(dna, [], fn c, acc -> [transcribe(c)|acc] end))
  end

  @spec transcribe(char) :: char
  defp transcribe(?C), do: ?G
  defp transcribe(?G), do: ?C
  defp transcribe(?T), do: ?A
  defp transcribe(?A), do: ?U
  
end
